neurodsp/rhythm/swm.py

```diff
@@ -149,17 +170,46 @@ def _compute_cost(sig, window_starts, win_n_samps):
     return cost
 
 
-def _find_new_window_idx(window_starts, spacing_n_samps, n_samp, tries_limit=1000):
-    """Find a new sample for the starting window."""
+def _find_new_window_idx(max_start, spacing_n_samps, current_window_idx, tries_limit=1000):
+    """Find a new sample for the starting window.
 
-    for n_try in range(tries_limit):
+    Parameters
+    ----------
+    max_start : int
+        The largest possible start index for the window.
+    spacing_n_samps : int
+        The minimum spacing required between windows, in samples.
+    current_window_idx : int
+        The current index of the window.
+    tries_limit : int, optional, default: False
+        The maximum number of iterations to attempt to find a new window.
 
-        # Generate a random sample & check how close it is to other window starts
-        new_samp = np.random.randint(n_samp)
-        dists = np.abs(window_starts - new_samp)
+    Returns
+    -------
+    new_samp : int
+        The index for the new window.
+    """
+
+    for _ in range(tries_limit):
+
+        # Find adjacent window starts
+        prev_idx = current_window_idx - (2*spacing_n_samps)
```

Comment:
my main concern is that this assumes the window positions based on their original positions, right? Which works in the first iteration, but as multiple windows have been moved, this seems like it would no longer guarantee non-overlap, right? Like, if the previous window has moved, checking the overlap based on it's original position might not be enough. 

So, I think this function should actually take in the full array of current window positions, and compare to that?

Reply:
As new positions are accepted, the new positions are updated before the next iteration. Lines 114-118 above:
```
if np.random.rand() < p_accept:

    # Update costs & windows
    costs[iter_num] = cost_temp
    window_starts = window_starts_temp
```

Reply:
yeh - but that's in the main function. The `_find_new_window_idx` doesn't use the `window_starts` variable, it only takes the current window, then infers the location of adjacent windows (lines 196-197), right? Within the context of `_find_new_window_idx`, it doesn't to have info about where the other windows actually are, I think? Am I completely missing something?

Reply:
Oh I see now! Yeah, I think we want what you described. I'll update and see how if the results change.

Reply:
Okay, so I updated `_find_new_window` to use actual window locations instead of inferring. However, the implementation can't find pure gaussians with random amounts of white noise between bursts (here's a [notebook](https://github.com/ryanhammonds/neurodsp/blob/swm/SWM.ipynb)). I'm lost as to why this is the case - it's probably a bug or a misunderstanding of the algorithm. 

Pre  367ebc2:
![pre](https://user-images.githubusercontent.com/34786005/105915751-edb24e80-5fe4-11eb-9af9-e76084a47140.png)

Post 367ebc2:
![post](https://user-images.githubusercontent.com/34786005/105915815-fe62c480-5fe4-11eb-93f6-cbfc70bd1012.png)

Maybe we should have a notebook with a few ideal cases to ensure the implementation here can do as well (or better) than the matlab version.

How SWM converges to a pattern is a bit confusing at it feels like a brute force hack that requires lots of computational power even with fancy (MCMC) resampling techniques. Even the matlab swm motif found in the notebook above is a a very messy pattern considering pure oscillations are simulated. Maybe we the new napp + bycycle approach could be a simpler alternative :wink:

Reply:
Hmm - that's a bit weird. In the simulated case, the "events" are right next to each other, so the algorithm should theoretically do better with a reduced value for `win_spacing` - does that help by any chance. 

Broadly - I don't want to get too pulled into a full audit of the method here. I think we should do diligence that (to the best of our knowledge) the algorithm proceeds as described in the paper, and is not obviously way worse than the matlab version. I think that's is broadly the case? (At least, the Matlab version in that notebook also doesn't do great on the noisy gaussian case). If you have a few minutes can you do a couple comparisons to matlab? If it looks comparable, I think we stick with this. 

I agree that how SWM converges is fairly brute force, and that TBH, it's probably not the best way to go - I would be somewhat more optimistic about spectral + temporal parameterization :)

Reply:
I missed your response here, sorry. I've ran the matlab version and the results are considerably better than after the algo update in 367ebc2. I've re-ran this code and what is currently here produces unstable results. Sometimes the waveform is a noisy gaussian and other times it's white noise. Reducing `win_spacing` didn't improve results either.

I'll have to dig into this deeper to see if there are any obvious bugs. The algorithm described in the paper seems relatively simple, but yet the matlab code is 2000+ lines long, and does a lot more than what is described in the algorithm section of the paper, so maybe I'm missing some other key part to the implementation.